### ISSUE TITLE
`Trainer(detect_anomaly=True)` not supported with `torch.compile`

### ISSUE THREAD

**Issue by awaelchli** [Lightning-AI/lightning#17591]
### Description & Motivation

The combination

```py
trainer = Trainer(detect_anomaly=True)
model = torch.compile(model)
trainer.fit(model)
```

is not supported by PyTorch. Issue: https://github.com/pytorch/pytorch/issues/100854

Currently, raises a compile error like this:

```
[/usr/local/lib/python3.10/dist-packages/torch/_subclasses/fake_tensor.py](https://localhost:8080/#) in local_scalar_dense(fake_mode, func, arg)
    461     if fake_mode.shape_env is None or not fake_mode.shape_env.allow_scalar_outputs:
    462         # Without symints/symfloats, cannot handle this
--> 463         raise DataDependentOutputException(func)
    464     if is_float_dtype(arg.dtype):
    465         return fake_mode.shape_env.create_unbacked_symfloat()

BackendCompilerFailed: backend='inductor' raised:
DataDependentOutputException: aten._local_scalar_dense.default
```

### Pitch

We could raise a more informative exception for the user early on. 

### Alternatives

Keep as is

cc @borda @justusschock @awaelchli @carmocca

**Comment by awaelchli** [comment]
This is fixed in pytorch 2.1.

**Comment by Borda** [comment]
> This is fixed in pytorch 2.1.

Maybe we can make better message with checking this option and pt version and compile?
